subsys/dfu/boot/mcuboot.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2016-2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_DOMAIN "fota/mcuboot"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FOTA_LEVEL
+#include <logging/sys_log.h>
+
+#include <stddef.h>
+#include <errno.h>
+#include <string.h>
+#include <flash.h>
+#include <zephyr.h>
+#include <init.h>
+
+#include "boot/mcuboot.h"
+#include "boot/mcuboot_constrains.h"
+
+/*
+ * Helpers for image trailer, as defined by mcuboot.
+ */
+
+#define TRAILER_IMAGE_MAGIC_SIZE (4 * sizeof(u32_t))
+#define TRAILER_SWAP_STATUS_SIZE (128 * FLASH_MIN_WRITE_SIZE * 3)
+#define TRAILER_COPY_DONE_SIZE FLASH_MIN_WRITE_SIZE
+#define TRAILER_IMAGE_OK_SIZE FLASH_MIN_WRITE_SIZE
+
+#define TRAILER_SIZE (TRAILER_IMAGE_MAGIC_SIZE + TRAILER_SWAP_STATUS_SIZE +\
+		      TRAILER_COPY_DONE_SIZE + TRAILER_IMAGE_OK_SIZE)
+
+#define TRAILER_PADDING	((uint8_t) BOOT_STATUS_ONGOING)
+
+static struct device *flash_dev;
+
+static u32_t boot_trailer(u32_t bank_offset)
+{
+	return bank_offset + FLASH_BANK_SIZE - TRAILER_SIZE;
+}
+
+static u32_t boot_trailer_magic(u32_t bank_offset)
+{
+	return boot_trailer(bank_offset);
+}
+
+static u32_t boot_trailer_swap_status(u32_t bank_offset)
+{
+	return boot_trailer_magic(bank_offset) + TRAILER_IMAGE_MAGIC_SIZE;
+}
+
+static u32_t boot_trailer_copy_done(u32_t bank_offset)
+{
+	return boot_trailer_swap_status(bank_offset) + TRAILER_SWAP_STATUS_SIZE;
+}
+
+static u32_t boot_trailer_image_ok(u32_t bank_offset)
+{
+	return boot_trailer_copy_done(bank_offset) + TRAILER_COPY_DONE_SIZE;
+}
+
+u8_t boot_status_read(void)
+{
+	u8_t img_ok = 0;
+	u32_t offset;
+
+	offset = boot_trailer_image_ok(FLASH_BANK0_OFFSET);
+	flash_read(flash_dev, offset, &img_ok, sizeof(u8_t));
+
+	return img_ok;
+}
+
+void boot_status_update(void)
+{
+	u8_t update_buf[TRAILER_IMAGE_OK_SIZE];
+	u32_t offset;
+	u8_t img_ok;
+
+	offset = boot_trailer_image_ok(FLASH_BANK0_OFFSET);
+	flash_read(flash_dev, offset, &img_ok, sizeof(u8_t));
+	if (img_ok == BOOT_STATUS_ONGOING) {
+		/* The first byte of the Image OK area contains payload. The */
+		/* rest is padded with 0xff for flash write alignment.       */
+		memset(update_buf, TRAILER_PADDING, sizeof(update_buf));
+		update_buf[0] = BOOT_STATUS_DONE;
+
+		flash_write_protection_set(flash_dev, false);
+		flash_write(flash_dev, offset, update_buf, sizeof(update_buf));
+		flash_write_protection_set(flash_dev, true);
+	}
+}
+
+void boot_trigger_image_update(void)
+{
+	u32_t copy_done_offset, image_ok_offset;
+	u8_t copy_done[TRAILER_COPY_DONE_SIZE];
+	u8_t image_ok[TRAILER_IMAGE_OK_SIZE];
+
+	copy_done_offset = boot_trailer_copy_done(FLASH_BANK1_OFFSET);
+	image_ok_offset = boot_trailer_image_ok(FLASH_BANK1_OFFSET);
+	memset(copy_done, TRAILER_PADDING, sizeof(copy_done));
+	memset(image_ok, TRAILER_PADDING, sizeof(image_ok));
```

Comment:
It shouldn't be necessary to write values that are already 0xFF.  Also, in the new code, these are defined to always be 8 bytes long, so that this code doesn't need to know the flash write alignment.